codegen/server/client_cli.go

```diff
@@ -0,0 +1,691 @@
+// parsing of the service subcommand.
+func BuildSubcommandData(svcName string, m *service.MethodData, buildFunction *BuildFunctionData, flags []*FlagData) *SubcommandData {
+	var (
+		name        string
+		fullName    string
+		description string
+
+		conversion string
+	)
+	{
+		en := m.Name
+		name = codegen.KebabCase(en)
+		fullName = goifyTerms(svcName, en)
+		description = m.Description
+		if description == "" {
+			description = fmt.Sprintf("Make request to the %q endpoint", m.Name)
+		}
+
+		if buildFunction == nil && len(flags) > 0 {
+			// No build function, just convert the arg to the body type
+			var convPre, convSuff string
+			target := "data"
+			if flagType(m.Payload) == "JSON" {
+				target = "val"
+				convPre = fmt.Sprintf("var val %s\n", m.Payload)
+				convSuff = "\ndata = val"
+			}
+			conv, check := conversionCode(
+				"*"+flags[0].FullName+"Flag",
+				target,
+				m.Payload,
+				false,
+			)
+			conversion = convPre + conv + convSuff
+			if check {
+				conversion = "var err error\n" + conversion
+				conversion += "\nif err != nil {\n"
+				if flagType(m.Payload) == "JSON" {
+					conversion += fmt.Sprintf(`return nil, nil, fmt.Errorf("invalid JSON for %s, example of valid JSON:\n%%s", %q)`,
+						flags[0].FullName+"Flag", flags[0].Example)
+				} else {
+					conversion += fmt.Sprintf(`return nil, nil, fmt.Errorf("invalid value for %s, must be %s")`,
+						flags[0].FullName+"Flag", flags[0].Type)
+				}
+				conversion += "\n}"
+			}
+		}
+	}
+	sub := &SubcommandData{
+		Name:          name,
+		FullName:      fullName,
+		Description:   description,
+		Flags:         flags,
+		MethodVarName: m.VarName,
+		BuildFunction: buildFunction,
+		Conversion:    conversion,
+	}
+	generateExample(sub, svcName)
+
+	return sub
+}
+
+// EndpointParserUsagesSection builds a section template that can be used to generate the endpoint usages code.
+func EndpointParserUsagesSection(data []*CommandData) *codegen.SectionTemplate {
+	usages := make([]string, len(data))
+	for i, cmd := range data {
+		subs := make([]string, len(cmd.Subcommands))
+		for i, s := range cmd.Subcommands {
+			subs[i] = s.Name
+		}
+		var lp, rp string
+		if len(subs) > 1 {
+			lp = "("
+			rp = ")"
+		}
+		usages[i] = fmt.Sprintf("%s %s%s%s", cmd.Name, lp, strings.Join(subs, "|"), rp)
+	}
+
+	return &codegen.SectionTemplate{Source: usageT, Data: usages}
+}
+
+// EndpointParserExamplesSection builds a section template that can be used to generate the endpoint examples code.
+func EndpointParserExamplesSection(data []*CommandData) *codegen.SectionTemplate {
+	var examples []string
+	for i, cmd := range data {
+		if i < 5 {
+			examples = append(examples, cmd.Example)
+		}
+	}
+
+	return &codegen.SectionTemplate{Source: exampleT, Data: examples}
+}
+
+// EndpointParserFlagsSection builds a section template that can be used to generate the endpoint flags code.
+func EndpointParserFlagsSection(data []*CommandData) *codegen.SectionTemplate {
```

Comment:
I would make this function simply return a `string` instead of a SectionTemplate. We can move the following code in gRPC and HTTP to this function and let it just return the `flagCode.String()`
```
  var flagsCode bytes.Buffer
  err := server.EndpointParserFlagsSection(data).Write(&flagsCode)
  if err != nil {
    panic(err)
  }
```
And let's call this function `FlagsCode` maybe?